monkey/monkey_island/cc/services/pth_report.py

```diff
@@ -0,0 +1,251 @@
+                    'machine': machine,
+                    'services': crit_machines[machine].get('critical_services'),
+                    'threatening_users': [i['name'] for i in crit_machines[machine]['threatening_users']]
+                }
+            )
+
+        return issues
+
+    @staticmethod
+    def get_strong_users_on_crit_details():
+        table_entries = []
+        user_details = {}
+        crit_machines = PTHReportService.get_strong_users_on_critical_machines_nodes()
+        for machine in crit_machines:
+            for user in crit_machines[machine]['threatening_users']:
+                username = user['name']
+                if username not in user_details:
+                    user_details[username] = {}
+                    user_details[username]['machines'] = []
+                    user_details[username]['services'] = []
+                user_details[username]['machines'].append(machine)
+                user_details[username]['services'] += crit_machines[machine]['critical_services']
+
+        for user in user_details:
+            table_entries.append(
+                {
+                    'username': user,
+                    'machines': user_details[user]['machines'],
+                    'services_names': user_details[user]['services']
+                }
+            )
+
+        return table_entries
+
+    @staticmethod
+    def generate_map_nodes():
+
+        nodes_list = []
+        monkeys = mongo.db.monkey.find({}, {'_id': 1, 'hostname': 1, 'critical_services': 1, 'ip_addresses': 1})
+        for monkey in monkeys:
+            critical_services = monkey.get('critical_services', [])
+            nodes_list.append({
+                'id': monkey['_id'],
+                'label': '{0} : {1}'.format(monkey['hostname'], monkey['ip_addresses'][0]),
+                'group': 'critical' if critical_services else 'normal',
+                'services': critical_services,
+                'hostname': monkey['hostname']
+            })
+
+        return nodes_list
+
+    @staticmethod
+    def generate_edge_nodes():
+        edges_list = []
+        pipeline = [
+            {
+                '$match': {'admin_on_machines': {'$ne': []}, 'secret_location': {'$ne': []}, 'type': 1}
+            },
+            {
+                '$project': {'admin_on_machines': 1, 'secret_location': 1}
+            }
+        ]
+        comp_users = mongo.db.groupsandusers.aggregate(pipeline)
+
+        for user in comp_users:
+            pairs = PTHReportService.generate_edges_tuples(user['admin_on_machines'], user['secret_location'])
+            for pair in pairs:
+                edges_list.append(
+                    {
+                        'from': pair[1],
+                        'to': pair[0],
+                        'id': str(uuid.uuid4())
+                    }
+                )
+        return edges_list
+
+    @staticmethod
+    def generate_edges_tuples(*lists):
```

Comment:
equiv to
[pair for pair in product(l1, l2) if pair[0] != pair[1]]
no need for function